packages/ember-runtime/lib/computed/reduce_computed_macros.js

```diff
@@ -5,18 +5,47 @@
 
 import Ember from 'ember-metal/core'; // Ember.assert
 import { get } from 'ember-metal/property_get';
-import {
-  guidFor
-} from 'ember-metal/utils';
 import EmberError from 'ember-metal/error';
-import run from 'ember-metal/run_loop';
-import { addObserver } from 'ember-metal/observer';
-import { arrayComputed } from 'ember-runtime/computed/array_computed';
-import { reduceComputed } from 'ember-runtime/computed/reduce_computed';
-import SubArray from 'ember-runtime/system/subarray';
+import { ComputedProperty, computed } from 'ember-metal/computed';
+import { addObserver, removeObserver } from 'ember-metal/observer';
 import compare from 'ember-runtime/compare';
+import { isArray } from 'ember-runtime/utils';
 
-var a_slice = [].slice;
+function reduceMacro(dependentKey, callback, initialValue) {
+  return computed(`${dependentKey}.[]`, function() {
+    return get(this, dependentKey).reduce(callback, initialValue);
+  }).readOnly();
+}
+
+function arrayMacro(dependentKey, callback) {
+  // This is a bit ugly
+  var fullKey = dependentKey;
+  if (/@each/.test(fullKey)) {
+    dependentKey = fullKey.replace(/\.@each.*$/, '');
```

Comment:
doesn't this change the invalidation behavior of the macro in a very unexpected way?